File: FloydWarshall.py
"""
 * FloydWarshall.py
 *
 * Implementation of Floyd Warshall Algorithm
 * 
 * See README for details
 * 
 * by:
 *    Ananth Murthy
 * 	  Chandan Yeshwanth
 * 
 * on:
 * 	  14th April 2014
 * 
"""

import sys 


# recursive function to obtain the path as a string
def obtainPath(i, j):
    if dist[i][j] == float("inf"):
        return " no path to "
    if parent[i][j] == i:
        return " "
    else :
        return obtainPath(i, parent[i][j]) + str(parent[i][j]+1) + obtainPath(parent[i][j], j)


if len(sys.argv) < 2:
	print "Check README for usage."
	sys.exit(-1)
	
try:	
	fil = open(sys.argv[1], "r")
except IOError:
	print "File not found."
	sys.exit(-1)


# no of vertices
V = int(fil.readline().strip())

# array of shortest path distances 
dist = []

# array of shortest paths
parent = []

# no of edges
E = int(fil.readline().strip())




# initialize to infinity
for i in range (0, V):
    dist.append([])
    parent.append([])
    for j in range (0, V):
        dist[i].append(float("inf"))
        parent[i].append(0)




# read edges from input file and store
for i in range (0,E):
    t = fil.readline().strip().split()
    x = int(t[0]) - 1
    y = int(t[1]) - 1
    w = int(t[2])
    dist[x][y] = w
    parent[x][y] = x


# path from vertex to itself is set to 0
for i in range (0,V):
    dist[i][i] = 0



# initialize the path matrix
for i in range (0,V):
    for j in range (0,V):
        if dist[i][j] == float("inf"):
            parent[i][j] = 0
        else:
            parent[i][j] = i



# actual floyd warshall algorithm
for k in range(0,V):
    for i in range(0,V):
        for j in range(0,V):
            if dist[i][j] > dist[i][k] + dist[k][j]:
                dist[i][j] = dist[i][k] + dist[k][j]
                parent[i][j] = parent[k][j]
                
                
                
# check for negative cycles
for i in range (0,V):
    if dist[i][i] != 0:
        print "Negative cycle at : ", i+1
        sys.exit() 


# display final paths
print "All Pairs Shortest Paths : \n"

# display shortest paths 
for i in range (0,V):
    print
    for j in range (0,V):
        print "From :", i+1, " To :", j+1
        print "Path :", str(i+1) + obtainPath(i,j) + str(j+1)
        print "Distance :", dist[i][j]
        print 

fil.close()

File: FloydWarshall.cc
/*
 * FloydWarshall.cc
 *
 * Implementation of Floyd Warshall Algorithm
 * 
 * Uses C++11 standard
 * 
 * See README for details
 * 
 * by:
 *    Ananth Murthy
 * 	  Chandan Yeshwanth
 * 
 * on:
 * 	  14th April 2014
 * 
 * */

# include <iostream>
# include <fstream>
# include <string>

using namespace::std;

// class to handle comparisions and additions of infinite weighted quantities, used to store distances
class Distance
{
private:

	int weight;
	bool infinite;
	
public:

	Distance()
	{
		weight = 0;
		infinite = true;
	}	
	
	void setWeight(int weight)
	{
		this->weight = weight;
		infinite = false;
	}
	
	string getWeight() const
	{
		if(infinite) 
			return "inf";
			
		return to_string(weight);
	}
	
	bool isInfinite()
	{
		return infinite;
	}
	
	bool isZero()
	{
		if(!infinite && weight==0)
			return true;
		
		return false;
	}
	
	// comparing infinite weights
	bool operator > (const Distance & e)
	{
		if (e.infinite)
			return false;
		else if (this->infinite)
			return true;
		else if (this->weight > e.weight)
			return true;
		
		return false;
	}
	
	// adding infinite weights
	Distance operator + (const Distance & e)
	{
		Distance d;
		
		if(this->infinite || e.infinite)
		{
			return d;
		}
		
		d.setWeight(e.weight + this->weight);
		return d;
	}
};



// To store shortest path distances
Distance ** dist ;

// To store the path and corresponding parents 
int ** parent ;



// Recursive function to obtain the path as a string
string obtainPath(int i, int j)
{
    if (dist[i][j].isInfinite())
        return " no path to ";
    
    if (parent[i][j] == i)
        return " ";
    else 
        return obtainPath(i, parent[i][j]) + to_string(parent[i][j]+1) + obtainPath(parent[i][j], j);
}



int main(int argc, char** argv)
{
	if(argc < 2)
	{
		cout << "Check README for usage." << endl;
		exit(-1);
	}
	
	ifstream ifile (argv[1]);
	
	if(!ifile)
	{
		cout << "File not found." << endl;
		exit(-1);
	}
	
	// number of vertices and edges
	int V, E;
	int i, j, k, u, v, w;
	
	ifile >> V >> E;
	
	
	// Matrices declared and initialised to infinity and zero respectively
	dist = new Distance * [V];
	for (int i = 0; i < V; i++)
		dist[i] = new Distance[V];
	
	
	parent = new int *[V];
	for (int i = 0; i < V; i++)
		parent[i] = new int[V];
	
	
	
	
	// Read edges from input file
	for (i=0; i<E; i++)
	{
		ifile >> u >> v >> w;
		dist[u-1][v-1].setWeight(w);
		parent[u-1][v-1] = u-1;
	}
	ifile.close();
	
	
	
	// Path from vertex to itself is set
	for (i=0; i<V; i++)
	{
		dist[i][i].setWeight(0);
		parent[i][i] = i;
	}
	
	
	// Actual Floyd Warshall Algorithm
	for (k=0; k<V; k++)
	{
		for (i=0; i<V; i++)
		{
			for (j=0; j<V; j++)
			{
				if (dist[i][j] > dist[i][k] + dist[k][j])
                {
					dist[i][j]    =   dist[i][k]    +   dist[k][j];
					parent[i][j]  =   parent[k][j];
				}
			}
		}
	}
	
	
	// Check for negative cycles
	for (i=0; i<V; i++)
	{
		if (!dist[i][i].isZero())
		{
			cout << "Negative cycle at : " << i+1 << endl;
			return 0;
		}
	}
	
	
	// Print all paths
	cout << "All Pairs Shortest Paths : \n\n";
	for (i=0; i<V; i++)
	{
		cout << endl;
		for (j=0; j<V; j++)
		{
			cout << "From : " << i+1 << " To : " << j+1 << endl;
			cout << "Path : " << 1+i << obtainPath(i,j) << j+1 << endl;
			cout << "Distance : " << dist[i][j].getWeight() << endl << endl;
		}
	}
	
	delete [] dist;
	delete [] parent;
	
	return 0;
}

